keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md

```diff
@@ -212,6 +211,17 @@ API server change:
     we will use the storage version objects to reconstruct a merged discovery
     document and serve that in all apiservers.
 
+Why so much work?
+* Note that merely serving 503s at the right times does not solve the problem,
+  for two reasons: controllers might get an incomplete discovery and therefore
+  not ask about all the correct resources; and when they get 503 responses,
+  although the controller can avoid doing something destructive, it also can't
+  make progress and is stuck for the duration of the upgrade.
```

Comment:
If a controller is trying to make progress it falls in one of two camps: 

1) it's trying to actuate on a resource which is removed in a newer version. In that case, whatever it does (besides deletions) are going to end up moot, because those objects will be orphaned once the upgrade completes.

2) it's trying to actuate on a resource which was just introduced in the newer version. In that case, if we are routing to the newest apiserver, this just actually does the right thing. 

So really, we're actually only talking about the first scenario and those resources are going to get orphaned, regardless of whether or not we end up merging the resources for discovery.

Reply:
I don't agree for case 1, the controller could be trying to delete the object in which case we want it to succeed.

Reply:
and anyway, we can't be sure the upgrade is going to complete.

Reply:
Then the correct thing to do is to route discovery to the newest apiserver and for resources which cannot be served by the newest apiserver, we should return a 404 instead of a 503, in order to force deletions.

Reply:
How does "the upgrade might not complete" imply that?

Reply:
I was responding mostly to this:

> I don't agree for case 1, the controller could be trying to delete the object in which case we want it to succeed.

But anyhow, it's cool, we chatted offline and I'm okay with merged discovery for beta. I just really wanted to avoid it for alpha.

Reply:
Well, that depends on the client. And during an upgrade, depending on ordering, we may see both clients that only understand v1beta1, and clients that only understand v1.